classworks/ivan-drobysh/15_websockets/demo/src/client/src/components/chat/chat.jsx

```diff
@@ -0,0 +1,141 @@
+import React from 'react';
+import { sendMessage, pushNewMessage } from 'resources/message/message.actions';
+import { getUsername } from 'resources/user/user.selectors';
+import { setRoomId } from 'resources/room/room.actions';
+import { getRoomId } from 'resources/room/room.selectors';
+import { loadMessages, deleteMessage, deleteMessageSelf } from 'resources/message/message.actions';
+import { getMessages } from 'resources/message/message.selectors';
+import { connect } from 'react-redux';
+import Message from './components/message/index';
+import io from 'socket.io-client';
+
+const socket = io();
+
+let timerId;
+import './chat.styles';
+
+class Chat extends React.Component {
+  onClick = (messageId) => () => {
+    return this.props.deleteMessage({ messageId });
+  }
+  componentWillMount() {
+    this.state = {
+      content: '',
+      roomId: this.props.roomId 
+    };
+
+    this.props.loadMessages({ roomId: this.state.roomId });
+    socket.emit('subscribe', { roomId: this.state.roomId || 'public'  });
+
+    socket.on('message:sent', message => {
+      this.props.pushNewMessage(message);
+    });
+
+    socket.on('message:remove', messageId => {
+      this.props.deleteMessageSelf({ messageId });
+    });
+
+    socket.on('typing', userId => {
+      if(userId)
+      this.setState({ isTyping: true, userId });
+      else
+        this.setState({ isTyping: false });
+    })
+  }
+  componentWillReceiveProps(nextProps) {
+    const { roomId } = nextProps;
+
+    if (this.state.roomId !== roomId) {
+      socket.emit('unsubscribe', { roomId: this.state.roomId || 'public' });
+
+      this.props.loadMessages({ roomId });
+
+      socket.emit('subscribe', { roomId: roomId || 'public' });
+    }
+
+    this.setState({ roomId });
+  }
+
+  sendMessage = async () => {
+    if (this.state.content) {
+      const message = { 
+        senderId: this.props.userId,
+        roomId: this.state.roomId,
+        content: this.state.content
+      };
+
+      await this.props.sendMessage(message);
+
+      // socket.emit('message:send', message, (res) => {
+      //   console.log('received response', res)
+      // });
+
+      this.setState({ content: '' });
+    }
+  };
+
+  onRoomChanged = (evt) => {
+    const roomId = evt.target.value || null;
+    this.props.setRoomId(roomId);
+    
+    return this.props.loadMessages({ roomId });
+  };
+
+  onChange = (e) => {
+    this.setState({ content: e.target.value });
+    socket.emit('typing', {
+      roomId: this.state.roomId,
+      userId: this.props.userId,
+    });
+    if(timerId)  clearTimeout(timerId);
+    timerId = setTimeout(()=> socket.emit('typing', {
```

Comment:
i don't get this logic. why do we send this event twice: with userId and without one. please describe